File: EasyUpiPayment/src/main/java/com/shreyaspatil/EasyUpiPayment/ui/PaymentUiActivity.java
package com.shreyaspatil.EasyUpiPayment.ui;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.shreyaspatil.EasyUpiPayment.R;
import com.shreyaspatil.EasyUpiPayment.Singleton;
import com.shreyaspatil.EasyUpiPayment.model.Payment;
import com.shreyaspatil.EasyUpiPayment.model.TransactionDetails;

import java.util.HashMap;
import java.util.Map;

public final class PaymentUiActivity extends AppCompatActivity {

    private static final String TAG = PaymentUiActivity.class.getSimpleName();

    public static final int PAYMENT_REQUEST = 4400;
    public static final String EXTRA_KEY_PAYMENT = "payment";

    private Singleton singleton;
    private Payment payment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_upipay);

        // Get instance of Singleton class
        singleton = Singleton.getInstance();

        // Get Payment Information
        Intent intent = getIntent();
        payment = (Payment) intent.getSerializableExtra(EXTRA_KEY_PAYMENT);

        // Set Parameters for UPI
        Uri.Builder payUri = new Uri.Builder();

        payUri.scheme("upi").authority("pay");
        payUri.appendQueryParameter("pa", payment.getVpa());
        payUri.appendQueryParameter("pn", payment.getName());
        payUri.appendQueryParameter("tid", payment.getTxnId());

        if (payment.getPayeeMerchantCode() != null) {
            payUri.appendQueryParameter("mc", payment.getPayeeMerchantCode());
        }

        payUri.appendQueryParameter("tr", payment.getTxnRefId());
        payUri.appendQueryParameter("tn", payment.getDescription());
        payUri.appendQueryParameter("am", payment.getAmount());
        payUri.appendQueryParameter("cu", payment.getCurrency());

        // Build URI
        Uri uri = payUri.build();

        // Set Data Intent
        Intent paymentIntent = new Intent(Intent.ACTION_VIEW);
        paymentIntent.setData(uri);

        // Check for Default package
        if (payment.getDefaultPackage() != null) {
            paymentIntent.setPackage(payment.getDefaultPackage());
        }

        // Show Dialog to user
        Intent appChooser = Intent.createChooser(paymentIntent, "Pay using");

        // Check if other UPI apps are exists or not.
        if (paymentIntent.resolveActivity(getPackageManager()) != null) {
            startActivityForResult(appChooser, PAYMENT_REQUEST);
        } else {
            Toast.makeText(this, "No UPI app found! Please Install to Proceed!",
                    Toast.LENGTH_SHORT).show();
            callbackOnAppNotFound();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == PAYMENT_REQUEST) {
            if (data != null) {
                // Get Response from activity intent
                String response = data.getStringExtra("response");

                if (response == null) {
                    callbackTransactionCancelled();
                    Log.d(TAG, "Payment Response is null");

                } else {
                    try {
                        // Get transactions details from response.
                        TransactionDetails transactionDetails = getTransactionDetails(response);

                        // Update Listener onTransactionCompleted()
                        callbackTransactionComplete(transactionDetails);

                        String status = transactionDetails.getStatus().toLowerCase();

                        // Check if success, submitted or failed
                        if (status.equals("success")) {
                            callbackTransactionSuccess();
                        } else if (status.equals("submitted")) {
                            callbackTransactionSubmitted();
                        } else {
                            callbackTransactionFailed();
                        }
                    } catch (Exception e) {
                        callbackTransactionCancelled();
                        callbackTransactionFailed();
                    }
                }
            } else {
                Log.e(TAG, "Intent Data is null. User cancelled");
                callbackTransactionCancelled();
            }
            finish();
        }
    }

    private Map<String, String> getQueryString(String url) {
        String[] params = url.split("&");
        Map<String, String> map = new HashMap<>();
        for (String param : params) {
            String name = param.split("=")[0];
            String value = param.split("=")[1];
            map.put(name, value);
        }
        return map;
    }

    // Make TransactionDetails object from response string
    private TransactionDetails getTransactionDetails(String response) {
        Map<String, String> map = getQueryString(response);

        String transactionId = map.get("txnId");
        String responseCode = map.get("responseCode");
        String approvalRefNo = map.get("ApprovalRefNo");
        String status = map.get("Status");
        String transactionRefId = map.get("txnRef");

        return new TransactionDetails(
                transactionId,
                responseCode,
                approvalRefNo,
                status,
                transactionRefId,
                payment.getAmount()
        );
    }

    // Checks whether listener is registered
    private boolean isListenerRegistered() {
        return (Singleton.getInstance().isListenerRegistered());
    }

    private void callbackOnAppNotFound() {
        Log.e(TAG, "No UPI app found on device.");

        if (isListenerRegistered()) {
            singleton.getListener().onAppNotFound();
        }
        finish();
    }

    private void callbackTransactionSuccess() {
        if (isListenerRegistered()) {
            singleton.getListener().onTransactionSuccess();
        }
    }

    private void callbackTransactionSubmitted() {
        if (isListenerRegistered()) {
            singleton.getListener().onTransactionSubmitted();
        }
    }

    private void callbackTransactionFailed() {
        if (isListenerRegistered()) {
            singleton.getListener().onTransactionFailed();
        }
    }

    private void callbackTransactionCancelled() {
        if (isListenerRegistered()) {
            singleton.getListener().onTransactionCancelled();
        }
    }

    private void callbackTransactionComplete(TransactionDetails transactionDetails) {
        if (isListenerRegistered()) {
            singleton.getListener().onTransactionCompleted(transactionDetails);
        }
    }
}

File: EasyUpiPayment/src/main/java/com/shreyaspatil/EasyUpiPayment/ui/PaymentUiActivity.kt
package com.shreyaspatil.easyupipayment.ui

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.shreyaspatil.easyupipayment.R
import com.shreyaspatil.easyupipayment.Singleton
import com.shreyaspatil.easyupipayment.exception.AppNotFoundException
import com.shreyaspatil.easyupipayment.model.Payment
import com.shreyaspatil.easyupipayment.model.TransactionDetails
import com.shreyaspatil.easyupipayment.model.TransactionStatus
import java.util.*

class PaymentUiActivity : AppCompatActivity() {

	private lateinit var payment: Payment

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_upipay)

		payment = (intent.getSerializableExtra(EXTRA_KEY_PAYMENT) as Payment?)
			?: throw IllegalStateException("Unable to parse payment details")

		// Set Parameters for UPI
		val paymentUri = Uri.Builder().apply {
			with(payment) {
				scheme("upi").authority("pay")
				appendQueryParameter("pa", vpa)
				appendQueryParameter("pn", name)
				appendQueryParameter("tid", txnId)
				appendQueryParameter("mc", payeeMerchantCode)
				appendQueryParameter("tr", txnRefId)
				appendQueryParameter("tn", description)
				appendQueryParameter("am", amount)
				appendQueryParameter("cu", currency)
			}
		}.build()

		// Set Data Intent
		val paymentIntent = Intent(Intent.ACTION_VIEW).apply {
			data = paymentUri

			// Check for Default package
			payment.defaultPackage?.let { `package` = it }
		}

		// Show Dialog to user
		val appChooser = Intent.createChooser(paymentIntent, "Pay using")

		// Check if other UPI apps are exists or not.
		if (paymentIntent.resolveActivity(packageManager) != null) {
			startActivityForResult(appChooser, PAYMENT_REQUEST)
		} else {
			Toast.makeText(
				this,
				"No UPI app found! Please Install to Proceed!",
				Toast.LENGTH_SHORT).show()
			throwOnAppNotFound()
		}
	}

	public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
		super.onActivityResult(requestCode, resultCode, data)
		if (requestCode == PAYMENT_REQUEST) {
			if (data != null) {
				// Get Response from activity intent
				val response = data.getStringExtra("response")

				if (response == null) {
					callbackTransactionCancelled()
					Log.d(TAG, "Payment Response is null")
				} else {
					runCatching {
						// Get transactions details from response.
						val transactionDetails = getTransactionDetails(response)

						// Update Listener onTransactionCompleted()
						callbackTransactionCompleted(transactionDetails)
					}.getOrElse {
						callbackTransactionCancelled()
					}
				}
			} else {
				Log.e(TAG, "Intent Data is null. User cancelled")
				callbackTransactionCancelled()
			}
			finish()
		}
	}

	// Make TransactionDetails object from response string
	@JvmSynthetic
	internal fun getTransactionDetails(response: String): TransactionDetails {
		return with(getMapFromQuery(response)) {
			TransactionDetails(
				transactionId = get("txnId"),
				responseCode = get("responseCode"),
				approvalRefNo = get("ApprovalRefNo"),
				transactionRefId = get("txnRef"),
				amount = payment.amount,
				transactionStatus = TransactionStatus.valueOf(
					get("Status")?.toUpperCase(Locale.getDefault())
						?: TransactionStatus.FAILURE.name
				)
			)
		}
	}

	@JvmSynthetic
	internal fun getMapFromQuery(queryString: String): Map<String, String> {
		val map = mutableMapOf<String, String>()
		val keyValuePairs = queryString
			.split("&")
			.map { param ->
				param.split("=").let { Pair(it[0], it[1]) }
			}
		map.putAll(keyValuePairs)
		return map
	}

	@JvmSynthetic
	internal fun throwOnAppNotFound() {
		Log.e(TAG, "No UPI app found on device.")
		throw AppNotFoundException(payment.defaultPackage)
	}

	@JvmSynthetic
	internal fun callbackTransactionCancelled() {
		Singleton.listener?.onTransactionCancelled()
	}

	@JvmSynthetic
	internal fun callbackTransactionCompleted(transactionDetails: TransactionDetails) {
		Singleton.listener?.onTransactionCompleted(transactionDetails)
	}

	companion object {
		const val TAG = "PaymentUiActivity"
		const val PAYMENT_REQUEST = 4400
		const val EXTRA_KEY_PAYMENT = "payment"
	}
}